test/test_autograd.py

```diff
@@ -5912,6 +5938,175 @@ def foo(a):
+
+            b = fwAD.make_dual(a, torch.rand(3))
+            self.assertFalse("tangent=" in str(a))
+            self.assertTrue("tangent=" in str(b))
+
+            b_primal, b_tangent = fwAD.unpack_dual(b)
+            self.assertFalse("tangent=" in str(b_primal))
+            self.assertFalse("tangent=" in str(b_tangent))
+
+    def test_grad_cleanup(self):
+        foo = torch.rand(2)
+        bar = torch.rand(2)
+        baz = torch.rand(2)
+
+        with fwAD.dual_level():
+            dual = fwAD.make_dual(foo, bar)
+            self.assertTrue("tangent=" in str(dual))
+
+        self.assertFalse("tangent=" in str(dual))
+
+        with fwAD.dual_level():
+            self.assertFalse("tangent=" in str(dual))
+            new_dual = fwAD.make_dual(foo, baz)
+
+            dual_primal, dual_tangent = fwAD.unpack_dual(dual)
+            new_dual_primal, new_dual_tangent = fwAD.unpack_dual(new_dual)
+            self.assertEqual(dual_primal, new_dual_primal)
+            self.assertIsNone(dual_tangent)
+            self.assertEqual(new_dual_tangent, baz)
+
+    def test_make_unpack_dual(self):
+        foo = torch.rand(2)
+        bar = torch.rand(2)
+
+        with fwAD.dual_level():
+            dual = fwAD.make_dual(foo, bar)
+
+            # dual should only be an alias to foo
+            self.assertEqual(dual.storage().data_ptr(), foo.storage().data_ptr())
+
+            # Make sure we don't break inplace checks for backward mode
+            self.assertEqual(foo._version, dual._version)
+            foo.add_(1)
+            self.assertEqual(foo._version, dual._version)
+
+            # Unpacking should only create aliases as well
+            dual_primal, dual_tangent = fwAD.unpack_dual(dual)
+            self.assertEqual(dual_primal.storage().data_ptr(), foo.storage().data_ptr())
+            self.assertEqual(dual_tangent.storage().data_ptr(), bar.storage().data_ptr())
+
+            # Make sure we don't break inplace checks for backward mode
+            self.assertEqual(foo._version, dual_primal._version)
+            foo.add_(1)
+            self.assertEqual(foo._version, dual_primal._version)
+            self.assertEqual(bar._version, dual_tangent._version)
+            bar.add_(1)
+            self.assertEqual(bar._version, dual_tangent._version)
+
+    def test_backward_mode(self):
+        foo = torch.rand(2, requires_grad=True)
+        bar = torch.rand(2, requires_grad=True)
+
+        with fwAD.dual_level():
+            dual = fwAD.make_dual(foo, bar)
+            p, t = fwAD.unpack_dual(dual)
+
+            gfoo, gbar = torch.autograd.grad(p.sum(), (foo, bar), retain_graph=True, allow_unused=True)
+            self.assertEqual(gfoo, torch.ones_like(foo))
+            self.assertIsNone(gbar)
+
+            gfoo, gbar = torch.autograd.grad(t.sum(), (foo, bar), retain_graph=True, allow_unused=True)
+            self.assertIsNone(gfoo)
+            self.assertEqual(gbar, torch.ones_like(bar))
+
+    def test_view(self):
```

Comment:
Actually a third option that is not discussed above is to have `make_dual()` do a full `.clone()` of the primal instead of a view. Maybe as a flag so that users that know that they won't do funky view of the make_dual's return can still have a no-copy version. Would that be ok?

Reply:
What's the "second option here"?

Reply:
The situation is even worse than you describe here. If you have:

```
x = ...
y0 = x[0:2]
y1 = x[1:3]
y1 += dual
```

you need y0 to also to become a dual tensor. This means that perturbations on view tensors must always be lazily computed based on the state of the base tensor at any given time.

Reply:
It changes the `handle_view` argument to `true` in `make_dual()` defined in `autograd.cpp`.
The impact is that the given Tensor to make_dual ends up being updated when we update the dual itself on the next line in `autograd.cpp`